src/gopher.cc

```diff
@@ -807,6 +813,10 @@ gopherReadReply(const Comm::ConnectionPointer &conn, char *buf, size_t len, Comm
 
         entry->timestampsSet();
         entry->flush();
+
+        if (!gopherState->len && !gopherState->overflowed)
```

Comment:
It is unclear what the first part of this condition means. A reader may be confused, assuming that gopherState->len means a kind of 'some bytes' received, which contradicts the statement below. Probably we should document gopherState->len (at least). It is used in gopherToHTML() - but I could not suggest a suitable definition based on its current usage.

Reply:
That looks like a branch bug to me, actually: Should not this be a positive `gopherState->len` test? 

```suggestion
        if (gopherState->len && !gopherState->overflowed)
```

If yes, that condition would match the "some bytes" in the corresponding markStoredReplyAsWhole() reason nicely. We could add something like `without overflowing` to that reason, but I think we do not have to document the _lack_ of problems in common/regular cases (or it may get out of hand because there are a lot of things that usually do _not_ go wrong.

If you agree that this is a bug, please reverse the condition. It would be a nice additional proof that these reason strings are useful!

Reply:
As we discussed, the '!gopherState->len' should work, i.e., we don't need to fix this line:
1. if we are on the gopherToHTML() path, non-zero len means an unterminated line received from the server (Gopher protocol requires all lines end with CRLF)
2. Otherwise, len remains zero (we do not look for CRLF).

However, I have some questions about it. For example, why do we consider a response 'complete' in (2) having the last line without a terminating CRLF? Also, the protocol requires '.' at the end of the message (i.e., messages are incomplete without it), but I read that many Gopher servers do not follow this rule.

Reply:
I think we are simply relying on the existing gopher response parsing code.

If that existing code considers the line complete, then we are not going to judge/improve it in this PR. I assume that `gopherState->len` is zero when that code considers the last line complete, but please correct me if I am wrong.

As for the trailing `.`, I think we should again follow the existing code lead: If that existing code is waiting for that period, then we ought to reflect that in our new condition. Otherwise, we are not going to overwrite that code decision. In other words, we simply leave that decision to that code, broken or not.

Reply:
> If that existing code considers the line complete

The old code does not consider an unterminated line complete (which is still in GopherStateData::buf) after receiving EOF, i.e., it does not add this line to the entry. On the other hand, the old code (upon receiving EOF) does not check that there is still buffered data (flushing it to the entry)  and treats the already received entry as 'complete', by calling both gopherEndHTML() and fwd->complete(). However, this is the place we are fixing here: we do not want to pretend that the entry is complete if it is not. Another approach would be to append the buffered (unterminated) line to the entry and declare the entry as 'complete' - but this would, even more, violate the Gopher specification (it is already violated by not looking for the terminating '.').

Reply:
> Alex: If that existing code considers the line complete

> Eduard: The old code does not consider an unterminated line complete

Yes, and that is fine:

* Existing parsing code considers the line complete if `gopherState->len` is zero. Our changes match that part of the algorithm.
* Existing parsing code considers the line incomplete if `gopherState->len` is positive. Our changes match that part of the algorithm.
* Existing entry completion code declares the entry complete regardless of `gopherState->len`. That is the bug we are fixing.

In summary, nothing to change here besides adding documentation for `len`, right?

Reply:
> nothing to change here besides adding documentation for len, right?

Agreed.

Reply:
I added that documentation at 867b855.

Reply:
BTW, `after receiving/storing some bytes` refers to the `!entry->isEmpty()` condition implied by the previous `if` statement. I think it is more-or-less decipherable from the surrounding code, but we can also add an "EOF after we stored (and, hence, received/parsed) some bytes" comment right after our `if` statement if you think it is needed.